xdem/volume.py

```diff
@@ -506,3 +507,194 @@ def local_hypsometric_interpolation(voided_ddem: Union[np.ndarray, np.ma.masked_
+    ddem_filled = ddem_arr.copy()
+    # Loop over all glaciers and fill the dDEM accordingly.
+    for i in unique_indices:
+        if i == 0:  # i==0 is assumed to mean stable ground.
+            continue
+        # Create a mask representing a particular glacier.
+        glacier_values = (glacier_index_map == i)
+
+        # The inlier mask is where that particular glacier is and where nans don't exist.
+        inlier_mask = glacier_values & ~ddem_nans
+
+        # If the fractional coverage is smaller than the given threshold, skip the glacier.
+        if (np.count_nonzero(inlier_mask) / np.count_nonzero(glacier_values)) < min_coverage:
+            continue
+
+        # Extract only the finite difference and elevation values that correspond to the glacier.
+        differences = ddem_arr[inlier_mask]
+        elevations = ref_arr[inlier_mask]
+
+        # Get the reference elevation min and max
+        elev_min = ref_arr[glacier_values].min()
+        elev_max = ref_arr[glacier_values].max()
+
+        # Copy the signal
+        signal = regional_signal["w_mean"].copy()
+        # Scale the signal elevation midpoints to the glacier elevation range.
+        midpoints = signal.index.mid
+        midpoints *= elev_max - elev_min
+        midpoints += elev_min
+        step = midpoints[1] - midpoints[0]
+        # Create an interval structure from the midpoints and the step size.
+        signal.index = pd.IntervalIndex.from_arrays(left=midpoints - step / 2, right=midpoints + step / 2)
+
+        # Find the hypsometric bins of the glacier.
+        hypsometric_bins = hypsometric_binning(
+            ddem=differences,
+            ref_dem=elevations,
+            bins=np.r_[[signal.index.left[0]], signal.index.right],  # This will generate the same steps as the signal.
+            kind="custom"
+        )
+        # Check which of the bins were non-empty.
+        non_empty_bins = np.isfinite(hypsometric_bins["value"])
+
+        # A theoretical minimum of two bins are needed for the rest of the analysis, but that's a bit low, right?
+        # I'll set this to 4 right now, and we can discuss this.
+        if np.count_nonzero(non_empty_bins) < 4:
+            continue
+
+        # Fit linear coefficients to scale the regional signal to the hypsometric bins properly.
+        # The inverse of the pixel counts are used as weights, to properly disregard poorly constrained bins.
+        coeffs = scipy.optimize.curve_fit(
+            f=lambda x, a, b: a * x + b,  # Estimate a linear function "f(x) = ax + b".
+            xdata=signal.values[non_empty_bins],  # The xdata is the normalized regional signal
+            ydata=hypsometric_bins["value"].values[non_empty_bins],  # The ydata is the actual values.
+            p0=[1, 0],  # The initial guess of a and b (doesn't matter too much)
+            sigma=1 / hypsometric_bins["count"].values[non_empty_bins]  # Weights are the inverse pixel count.
+        )[0]
```

Comment:
interesting approach! I hadn't thought of doing a linear fit between the glacier-wise signal and the regional signal for scaling the regional signal like this. I'm not sure I fully comprehend the behavior but I expect it to perform quite well!